### DISCUSSION TITLE
Add transparency to image with cpp

### DISCUSSION THREAD

**Discussion by jackthenewbie** [ImageMagick/ImageMagick#6191] [Help] [answered]
### ImageMagick version

7

### Operating system, version and so on

ubuntu-20

### Description

https://stackoverflow.com/questions/35554640/how-to-add-transparency-to-a-png-image-with-imagemagick
I'm having some trouble to add transparency to my image, I can't figure it out how to put this command into cpp code.

**Comment by snibgo** [comment]
What have you tried so far?

**Reply by jackthenewbie** [reply]
```cpp
        Image artifact("artifact.png");
        artifact.resize(Geometry(icon_size, icon_size));
        Image mask(artifact);
        mask.channel(OpacityChannel);
        mask.colorFuzz(QuantumRange * 0.5);
        mask.opaque(Color("black"), Color("gray50"));
        mask.negate();
        artifact.composite( mask, 0, 0, CopyAlphaCompositeOp );
        this->drawList.push_back(DrawableCompositeImage(ximg, yimg, 0, 0, artifact, OverCompositeOp));
```
Honestly, I don't even know what im doing, [this](https://stackoverflow.com/questions/30534626/how-to-set-opacity-to-png-image-with-transparent-background-using-magick) is the current one that I copy. But the result not as I want.

**Comment by snibgo** [comment]
> But the result not as I want.

Okay, you have shown examples that don't do what you want. What do you want, exactly?

**Reply by jackthenewbie** [reply]
![image](https://user-images.githubusercontent.com/55600733/227666021-7fce90e4-4831-4277-a77c-1d5a7e6f6334.png)
Apparently, I got this with the code above, and what I want is [this](https://stackoverflow.com/questions/35554640/how-to-add-transparency-to-a-png-image-with-imagemagick)

**Comment by snibgo** [comment, marked as answer]
The link has the CLI for multiplying the alpha channel by 0.7. So:
```
    image.alpha (true);
    image.evaluate (AlphaChannel, MultiplyEvaluateOperator, 0.7);
```
Does that do what you want?

**Reply by jackthenewbie** [reply]
yup it works, may I ask how do you know this, I'm so confuse where to start, can't understand docs at all

**Comment by snibgo** [comment]
The main source code for the CPP interface is `Magick++\lib\Image.cpp`. With a text editor I searched for the word `evaluate`. The first two arguments are of type `ChannelType` and `MagickEvaluateOperator`. I searched for those in the `*.h` files to find suitable values. Those types are declared in `pixel.h` and `statistic.h`.

So that takes care of multiplying the alpha channel. But some images are fully opaque and have no alpha channel. Another search of `Magick++\lib\Image.cpp` quickly found how to ensure an alpha channel exists.
